### ISSUE TITLE
let-union_type is narrowed incorrectly

### ISSUE THREAD

**Issue by gabor** [microsoft/TypeScript#31580]
**TypeScript Version:**  3.4.0-dev.20190524

**Search Terms:**
union narrow return
**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
function getThing(): string {
    let result:string|number = 'text';

    const setItTo42 = () => {
        result = 42;
    }

    setItTo42();

    return result;
}
```

**Expected behavior:**
an error message describing that i cannot return the potentially-number value from a function that should return a string.

**Actual behavior:**
no errors are shown.

**Playground Link:** 

**Related Issues:** 

please note, that the following code produces the correct result (an error-message is shown):
```ts
function getThing(): string {
    let result:string|number = 'text';

    (() => {
      result = 42;
    })();

    return result;
}
```
the only difference being that in one case the modifier-function is assigned to a variable and then called, and in the other case it is immediately called.

**Comment by MartinJohns** [comment]
Duplicate of #9998.

**Comment by fatcerberus** [comment]
Indeed yet another duplicate of #9998, however this one is interesting since it's the first time I've heard of the current optimistic behavior leading to something unsound.  Usually people complain about the narrowings *not* being discarded! :smile:
